### ISSUE TITLE
Add real user with home directory

### ISSUE THREAD

**Issue by bgruening** [bgruening/docker-ipython-notebook#28]
Instead of dropping privileges with `USER nobody` we could have a real user that can use `pip` and `install.packages()`.

**Comment by erasche** [comment]
+1 for not using nobody.
6. mar. 2015 13.36 skrev "Björn Grüning" <notifications@github.com>:

**Comment by bgruening** [comment]
I think the cleanest solution is to add the same user (uid/gid) as galaxy-stable to this container. At least until this is fixed: https://github.com/docker/docker/pull/11253.

**Comment by erasche** [comment]
Yep, sounds good. Did we not try this before?

tir. 10. mar. 2015, 18.42 skrev Björn Grüning <notifications@github.com>:

**Comment by bgruening** [comment]
@erasche we had a user before, but this was not the same user as in the parent Galaxy container. With the applied change this should hopefully work.

**Comment by jmchilton** [comment]
Is a better work around to leave the Docker container to run as root, read the the uid on ``/import``, create the IPython user at runtime startup with that UID, and then su into that new user to run the IPython server?

Potentially we could even pre-create everything the way you down now with a fixed UID of 1450 and just skip that logic if the UID is 1450 - and just chown everything all at once very efficiently if you do have to create a new user in-container.

**Comment by erasche** [comment]
Yes, we could get away with `USER root` and dropping privileges on boot for the ipython process only. That would get around the uid/gid issue with the host.

That said, I'm not a big fan of that, because it means more IE specific logic that has to be encoded into the boot steps of every IE, but I don't have an alternative either.

**Comment by jmchilton** [comment]
Other IEs may not need a ``$HOME`` right - this may be sort of specific to IPython and RStudio?

Regardless though - on balance isn't pointing more logic in the container better than requiring Galaxy to have a specific UID?

If the answer to that is no - maybe the approach shouldn't be that we provide fixed containers for people but maybe instead provide Dockerfiles with onbuild instructions or something to allow people to tailor the container to their deployment?

**Comment by erasche** [comment]
Hmm, fair point, it may be more specific to IEs with longer lasting workspaces. I worry that there will be a lot of those (codebox comes to mind, jbrowse/webapollo run as nginx and tomcat respectively, so there shouldn't be issues there), but that's getting ahead of ourselves.

No, anything is better than requiring Galaxy to have a specific UID, that will never be an acceptable solution. 

True, we could have Galaxy build containers (or thin wrappers over existing IEs) but that feels like murky territory from "download this image at this tag, run it, done". 

So I guess you've swayed me, I'd rather IE coded logic than anything outside of it. :)

**Comment by bgruening** [comment]
The proper solution I had hoped for was postponed to docker 1.7 a few hours ago :( https://github.com/docker/docker/pull/11253

Until user namespace support is in docker we should go with the second suggestion from John, imho. That is: detect if the UID is different than 1450 and chmod everything at startup.

**Comment by erasche** [comment]
@jmchilton 

> Potentially we could even pre-create everything the way you down now with a fixed UID of 1450

galaxy create files with a different UID? Will this not require a `su` exemption for doing the `chown` before starting the IE?

@bgruening we're `USER ipython`, how do you plan to chown in the container? 
It's unfortunate that that issue was postponed, as that really would be the better option.

**Comment by jmchilton** [comment]
I meant pre-create `/home/ipython` in the Docker container as part of the Dockerfile instructions (build time) - not chown stuff outside the container (run time).

**Comment by erasche** [comment]
Ah, thanks for the clarity. Don't we do that already? https://github.com/bgruening/docker-ipython-notebook/blob/master/Dockerfile#L30

**Comment by jmchilton** [comment]
To clarify what I was trying to say...

 - Do everything we are doing now at build time, except Switch ``USER root``
 - On startup add logic so:
    - If uid of ``/import`` ``== 1450`` ``pass``
    - Else - delete ``ipython`` user,  ``chown -R`` ``/home/ipython`` with that uid and create a new ``ipython`` user with this uid.
   - Use ``su`` to run ``ipython`` as ``ipython``.

**Comment by erasche** [comment]
Ah, sorry, given the other discussion it sounded like @bgruening meant just the second part separately, and I was confused. 

This resolves that but gets back to `USER root` which is something we'd wanted to avoid, but it definitely makes sense in the grand scheme of things until the other mentioned issue can be implemented.

**Comment by jmchilton** [comment]
Yeah - indeed - I too would like to avoid that ideally. I wonder if there are any security benefits of specifying a different user versus starting with ``USER root`` and then dropping permissions. Probably these are comparable from a security perspective as long as there is no way to "inject" stuff to the commands that root is running.

We could also probably avoid ``USER root`` with an effective uid script - a little benefit but not tons I guess.

**Comment by erasche** [comment]
All sounds good. We don't live in an ideal world, so we may have to use `root` for a while until docker gets itself sorted out. I'm now wholly in favour of all of this as it sounds like the best option for dealing with ID conflicts.
